File: week_8_and_9/3_JavaScript/grocery_list/my_solution.js
// U3.W8-9: 


// I worked on this challenge [by myself].

// 2. Pseudocode

// 1. Create a grocery constructor function
// 2. Make it so you can view the grocery list by utilizing objects
// 3. Make it so you can add items in the constructor function
// 4. Make it so you can delete items inside the grocery list
// 5. Create a function that counts the amount of items inside of the grocery list
// 6. Create a function that will count the amount of specific items in the grocery list


// 3. Initial Solution

// itemArray = ["apple", "banana", "orange"];

// numberOfItems = [3, 6, 4];

// var groceryList = {
//   addItem: function(item_to_add, number)
//   {
// 		for (var i = 0; i < itemArray.length; i++)
// 		{
// 			itemArray.push(item_to_add);
// 			return numberOfItems.push(number);
// 		}
// 	},
	
// 	viewGroceryList: function()
// 	{
// 		for (var i = 0; i < itemArray.length; i++)
// 		{
// 			console.log(itemArray[i])
// 			console.log(numberOfItems[i])
// 		}
// 	},
	
// 	amountOfGroceries: function() 
// 	{
// 	return console.log("There are " + itemArray.length + " groceries in the list.")
// 	},

// 	groceryCount: function(item)
// 	{
// 		var i = itemArray.indexOf(item)
// 		return console.log(numberOfItems[i])
// 	},

// 	removeItem: function(item)
// 	{
// 		var i = itemArray.indexOf(item)
// 		itemArray.splice(i, 1)
// 		numberOfItems.splice(i, 1)
// 	}
// };

// 4. Refactored Solution

itemArray = ["apple", "banana", "orange"];

numberOfItems = [3, 6, 4];

var groceryList = {
	
  addItem: function(item_to_add, number)
  {
		for (var i = 0; i < itemArray.length; i++)
		{
			itemArray.push(item_to_add);
			return numberOfItems.push(number);
		}
	},
	amountOfGroceries: function() 
	{
	return console.log("There are " + itemArray.length + " groceries in the list.")
	},

	groceryCount: function(item)
	{
		var i = itemArray.indexOf(item)
		return console.log(numberOfItems[i])
	},

	removeItem: function(item)
	{
		var i = itemArray.indexOf(item)
		itemArray.splice(i, 1)
		numberOfItems.splice(i, 1)
	},

	viewGroceryList: function()
	{
		for (var i = 0; i < itemArray.length; i++)
		{
			console.log(itemArray[i])
			console.log(numberOfItems[i])
		}
	}
	
};




// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

function assert(test, message, test_number) {
  if (!test) {
    console.log(test_number + "false");
    throw "ERROR: " + message;
  }
  console.log(test_number + "true");
  return true;
}

//groceryList.addItem("bacon", 3); - works

// groceryList.removeItem("banana") - works

// groceryList.viewGroceryList() - works

//groceryList.amountOfGroceries() - works

//groceryList.groceryCount("banana") - works

assert(
	(groceryList.addItem instanceof Function),
	"Should be a function\n",
	"1. ")

assert(
	(groceryList.viewGroceryList instanceof Function),
	"Should be a function\n",
	"2. ")

assert(
	(groceryList.amountOfGroceries instanceof Function),
	"Should be a function\n",
	"3. ")

assert(
	(groceryList.groceryCount instanceof Function),
	"Should be a function\n",
	"4. ")

assert(
	(groceryList.removeItem instanceof Function),
	"Should be a function\n",
	"5. ")

// 5. Reflection 

// What parts of your strategy worked? What problems did you face?
// I'm having issues with the driver code, I was unable to make the values come out as true, even though they were still equal to eachother.
// I did make sure the code worked as I was coding with node.js, however I do want further education on this matter. I'll do some extra research on my own where there is some spare time to do so.
// Everything was pretty simple, since my practice with JS is limited, I did what I could to make everything as DRY as possible, all of my code works the way it should.

// What questions did you have while coding? What resources did you find to help you answer them?
// I learned the splice method, it was very useful.

// What concepts are you having trouble with, or did you just figure something out? If so, what?
	// Javascript is something I need to work more with.

// Did you learn any new skills or tricks?
	// I learned how to splice throuugh arrays

// How confident are you with each of the Learning Competencies?
	// Not too confident, but I feel that if I practice more with JS, I'll be good to go

// Which parts of the challenge did you enjoy?
	//It was honestly harder for me to work with, but I enjoyed the refresher.

// Which parts of the challenge did you find tedious?
	//I felt like I got stuck with the driver code. Even without the console.log statements inside of my program, I was unable to get my code to run through assert methods the way I wanted, but it all ran the way I wanted.
	//This is something I'll have to learn more about in the very near future.

File: week_8_and_9/4_Ruby/triangle_debugger/my_solution.rb
# U3.W8-9: Triangle Debugger


# I worked on this challenge [by myself, with: ].


# 1. Original Solution

# def valid_triangle?(a, b, c, sum)
#   if a != 0 || b != 0 || c != 0
#   if a >= b
#   largest = a
#   sum += b
#   else largest = b
#   sum += a
#   end
#   if c > largest
#   sum += largest
#   largest = c
#   else sum += c
#   end
#   if sum > largest
#   return "true"
#   else return "false"
#   end
#   else return "false"
#   end
# end

def valid_triangle?(a, b, c)
  sum = 0
  return false if a <= 0 || b <= 0 || c <= 0
  return false if a + b < c || a + c < b || b + c < a
  if a != 0 || b != 0 || c != 0
    if a >= b
      largest = a
      sum += b
    else largest = b
      sum += a
    end
    if c > largest
      sum += largest
      largest = c
    else 
      sum += c
    end
    if sum > largest
      return "true"
    else 
      return "false"
    end
  else 
    return "false"
  end
end


# 2. Answer the question for each bug

# --- Bug 1 ---
# * what is the exact description of the error?
#   valid_triangle? expects 3 arguments
# * what is the exact line number the error is appearing?
#    line 9
# * before you fix the bug, what do you think is causing the error?
#   the sum arguement 

# --- Bug 2 ---
# * what is the exact description of the error?
#   valid_triangle? returns false if any of the arguments are 0
# * what is the exact line number the error is appearing?
#   line 20
# * before you fix the bug, what do you think is causing the error?

# --- Bug 3 ---
# * what is the exact description of the error?
#   valid_triangle? returns false for any impossible triangle
# * what is the exact line number the error is appearing?
#   line 46 in spec file (not returning expected value)
# * before you fix the bug, what do you think is causing the error?
#   It's because we don't have any test to check for an invalid traingle (a + b > c)

# 3. Refactored Solution (Comment the other code to run this)

def valid_triangle?(a, b, c)
  a + b > c && a + c > b && b + c > a ? true : false
end


# 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

def assert
  raise "assertion failed" unless yield 
end

assert { valid_triangle?(0,2,2) == false }
assert { valid_triangle?(2,2,2) == true }
assert { valid_triangle?(1,2,3) == false }


# U3.W8-9: Triangle Debugger


# I worked on this challenge [by myself, with: ].


# 1. Original Solution

# def valid_triangle?(a, b, c, sum)
#   if a != 0 || b != 0 || c != 0
#   if a >= b
#   largest = a
#   sum += b
#   else largest = b
#   sum += a
#   end
#   if c > largest
#   sum += largest
#   largest = c
#   else sum += c
#   end
#   if sum > largest
#   return "true"
#   else return "false"
#   end
#   else return "false"
#   end
# end

def valid_triangle?(a, b, c)
  sum = 0
  return false if a <= 0 || b <= 0 || c <= 0
  return false if a + b < c || a + c < b || b + c < a
  if a != 0 || b != 0 || c != 0
    if a >= b
      largest = a
      sum += b
    else largest = b
      sum += a
    end
    if c > largest
      sum += largest
      largest = c
    else 
      sum += c
    end
    if sum > largest
      return "true"
    else 
      return "false"
    end
  else 
    return "false"
  end
end


# 2. Answer the question for each bug

# --- Bug 1 ---
# * what is the exact description of the error?
#   valid_triangle? expects 3 arguments
# * what is the exact line number the error is appearing?
#    line 9
# * before you fix the bug, what do you think is causing the error?
#   the sum arguement 

# --- Bug 2 ---
# * what is the exact description of the error?
#   valid_triangle? returns false if any of the arguments are 0
# * what is the exact line number the error is appearing?
#   line 20
# * before you fix the bug, what do you think is causing the error?

# --- Bug 3 ---
# * what is the exact description of the error?
#   valid_triangle? returns false for any impossible triangle
# * what is the exact line number the error is appearing?
#   line 46 in spec file (not returning expected value)
# * before you fix the bug, what do you think is causing the error?
#   It's because we don't have any test to check for an invalid traingle (a + b > c)

# 3. Refactored Solution (Comment the other code to run this)

def valid_triangle?(a, b, c)
  a + b > c && a + c > b && b + c > a ? true : false
end


# 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE

def assert
  raise "assertion failed" unless yield 
end

assert { valid_triangle?(0,2,2) == false }
assert { valid_triangle?(2,2,2) == true }
assert { valid_triangle?(1,2,3) == false }

# 5. Reflection 

# What parts of your strategy worked? What problems did you face?
#We had an easy time with this challenge. It was fun. We realized that if you put 0 anywhere in our code it would come out to being false, so we only needed one line of code.

# What questions did you have while coding? What resources did you find to help you answer them?
  #Getting arity to work with methods is different from classes, that's something I'll need to research.

# What concepts are you having trouble with, or did you just figure something out? If so, what?
  #Not really any, I just need to review some things.

# Did you learn any new skills or tricks?
  #Nope

# How confident are you with each of the Learning Competencies?
  #Pretty confident, ruby isn't hard to work with

# Which parts of the challenge did you enjoy?
  #All really, I love working with ruby

# Which parts of the challenge did you find tedious?
  #none